File: StrandiOSShared/HrTrace.swift
import Foundation

/// One point on the widget's heart-rate trace: when it was taken, and the bpm.
///
/// `Sendable` because a SwiftUI `Shape` holding these is itself Sendable, and the compiler warns that
/// carrying a non-Sendable stored property there is an error in a future language mode. Two immutable
/// value fields, so it is Sendable by inspection rather than by assertion.
public struct HrPoint: Codable, Equatable, Sendable {
    public let ts: Int64
    public let bpm: Int
    public init(ts: Int64, bpm: Int) {
        self.ts = ts
        self.bpm = bpm
    }
}

/// The pure half of the heart-rate widget: what the trace CONTAINS and where its ink goes.
///
/// Swift twin of the Kotlin `HrTrace`. The retention rule, the tick choices and the normalisation are
/// the same decisions on both platforms and must not be made twice — a trace that kept a different
/// window, or picked different labels, would be a different reading of the same heart.
///
/// What deliberately does NOT port is the bitmap sizing: `fitBox`, the byte budget and the 565 pixel
/// depth exist because Glance compiles to RemoteViews, cannot draw, and carries its payload over a
/// Binder transaction with a hard ceiling. WidgetKit is SwiftUI and strokes the trace as a vector, so
/// there is no bitmap to budget for and nothing here needs a pixel format.
///
/// `encode`/`decode` do not port either. Android packs the series into a SharedPreferences string
/// because that is what it has; the snapshot here is already Codable and carries `[HrPoint]` directly,
/// so a hand-rolled format would be a second encoding to keep correct for no gain.
public enum HrTrace {

    /// How much history the trace shows. Three hours matches what fits legibly at widget width.
    public static let windowSec: Int64 = 3 * 60 * 60

    /// One point per minute. The strap streams ~1/s; keeping every sample would put 10 800 points
    /// behind a trace a few hundred points wide, which costs space and buys nothing the eye resolves.
    public static let bucketSec: Int64 = 60

    /// What counts as a GAP. Deliberately NOT the Today chart's "more than one bucket": that chart reads
    /// a fixed DB grid, where a missing bucket really is missing data, while this series only gains a
    /// point when the app PUBLISHES one, and background scheduling routinely skips a minute. At a
    /// one-bucket threshold that ordinary jitter would shatter a healthy trace into dots, which is a
    /// worse lie than the joined line being fixed.
    ///
    /// Set to `HrDisplay.staleCap` rather than to a number picked here, so the line breaks exactly where
    /// the widget would already have dropped the headline reading for being too old to represent HR at
    /// all. `testTheGapThresholdMatchesTheStaleCap` pins the two together.
    public static let gapSec: Int64 = 15 * bucketSec

    /// Hard cap, so a clock jump backwards cannot grow the series without bound.
    public static let maxPoints = 200

    /// Fold a fresh reading into the series: one point per minute bucket, newest wins within a bucket.
    ///
    /// Newest-wins rather than an average because the widget's headline number is the LATEST bpm, and a
    /// trace whose last point disagreed with the number printed above it would read as a bug.
    public static func append(_ series: [HrPoint], ts: Int64, bpm: Int, nowSec: Int64? = nil) -> [HrPoint] {
        let now = nowSec ?? ts
        guard bpm > 0 else { return prune(series, nowSec: now) }
        let bucket = ts / bucketSec * bucketSec
        var out = series.filter { $0.ts != bucket }
        out.append(HrPoint(ts: bucket, bpm: bpm))
        // No sort here: `prune` sorts on the way out and has to anyway, since it is also the entry point
        // for a decoded series. Appending at the end is already in order for anything but a backwards
        // clock, and prune's sort is what makes even that correct.
        return prune(out, nowSec: now)
    }

    /// Drop anything older than the window, then anything beyond the cap (oldest first).
    public static func prune(_ series: [HrPoint], nowSec: Int64) -> [HrPoint] {
        let floor = nowSec - windowSec
        let kept = series.filter { $0.ts >= floor }.sorted { $0.ts < $1.ts }
        guard kept.count > maxPoints else { return kept }
        return Array(kept.suffix(maxPoints))
    }

    /// The three numbers the header shows. Nil when there is nothing to describe.
    public struct Stats: Equatable {
        public let min: Int
        public let max: Int
        public let latest: Int
    }

    public static func stats(_ series: [HrPoint]) -> Stats? {
        guard let first = series.first, let last = series.last else { return nil }
        var lo = first.bpm
        var hi = first.bpm
        for p in series {
            if p.bpm < lo { lo = p.bpm }
            if p.bpm > hi { hi = p.bpm }
        }
        return Stats(min: lo, max: hi, latest: last.bpm)
    }

    /// A point in the trace's drawing box, origin top-left.
    ///
    /// `startsRun` means LIFT THE PEN before this point: nothing was recorded for `gapSec` before it.
    /// The trace joined every point unconditionally, and because x is mapped by TIME rather than by
    /// index, a 90-minute disconnect inside the 3-hour window drew as one confident diagonal across
    /// half the widget. The gap was already the right width; it was the line across it that was never
    /// measured: the same defect #2082 describes on the Today sparkline, on a second surface.
    public struct Pt: Equatable {
        public let x: CGFloat
        public let y: CGFloat
        public let startsRun: Bool

        public init(x: CGFloat, y: CGFloat, startsRun: Bool = false) {
            self.x = x
            self.y = y
            self.startsRun = startsRun
        }
    }

    /// The trace split into runs of consecutive readings: each range is a stretch the strap recorded
    /// without a break, and the pen lifts between one run and the next. A series with no gaps is one
    /// run, which is the common case and draws exactly as it always did.
    public static func runs(_ points: [Pt]) -> [ClosedRange<Int>] {
        guard !points.isEmpty else { return [] }
        var out: [ClosedRange<Int>] = []
        var start = 0
        for i in 1..<points.count where points[i].startsRun {
            out.append(start...(i - 1))
            start = i
        }
        out.append(start...(points.count - 1))
        return out
    }

    /// The trace as coordinates inside a `width` x `height` box.
    ///
    /// Two degenerate shapes decide most of this, and both are ordinary rather than exotic — a widget
    /// placed mid-afternoon has one point, and a resting arm holds one bpm for minutes at a time:
    ///
    ///  - ONE point, or every point at the same instant: there is no time axis to spread across, so it
    ///    sits at the left edge rather than at x = NaN.
    ///  - every bpm equal: there is no range to scale into, so the line runs along the VERTICAL MIDDLE.
    ///    Pinning it to the top or bottom would read as a maxed-out or flatlined heart.
    ///
    /// Y is flipped on the way out: bpm rises upward, screen coordinates rise downward.
    public static func points(_ series: [HrPoint], width: CGFloat, height: CGFloat) -> [Pt] {
        guard !series.isEmpty, width > 0, height > 0,
              let t0 = series.first?.ts, let t1 = series.last?.ts else { return [] }
        let span = CGFloat(t1 - t0)
        var lo = series[0].bpm
        var hi = series[0].bpm
        for p in series {
            if p.bpm < lo { lo = p.bpm }
            if p.bpm > hi { hi = p.bpm }
        }
        let range = CGFloat(hi - lo)
        return series.enumerated().map { i, p in
            let x = span <= 0 ? 0 : CGFloat(p.ts - t0) / span * width
            let y = range <= 0 ? height / 2 : height - CGFloat(p.bpm - lo) / range * height
            return Pt(x: x, y: y, startsRun: i > 0 && p.ts - series[i - 1].ts > gapSec)
        }
    }

    /// The three bpm labels down the right edge: max, midpoint, min, stacked as they appear.
    ///
    /// A flat series returns the same number three times rather than an invented spread — the honest
    /// reading of a steady heart is one number. The VIEW declines to draw a scale in that case; the
    /// rule stays here so both platforms agree on what the labels would be.
    public static func bpmTicks(_ stats: Stats) -> [Int] {
        [stats.max, (stats.min + stats.max + 1) / 2, stats.min]
    }

    /// The three timestamps along the bottom: first, middle, last of the DATA, not of the window.
    ///
    /// Anchored to the data because a widget holding forty minutes of history would otherwise label its
    /// axis with two hours that were never sampled. Fewer than three distinct instants returns what
    /// there is, so the view draws one label rather than three copies of it.
    public static func timeTicks(_ series: [HrPoint]) -> [Int64] {
        guard let first = series.first?.ts, let last = series.last?.ts else { return [] }
        if first == last { return [first] }
        let mid = first + (last - first) / 2
        if mid == first || mid == last { return [first, last] }
        return [first, mid, last]
    }
}

/// What heart rate the widget should SHOW, given how long ago it was actually measured.
///
/// Twin of the Kotlin `HrDisplay`, and the reason it exists is the same on both platforms: a snapshot
/// carries the last reading indefinitely, so without an age rule a widget renders an hours-old number as
/// though it were current.
///
/// Anchored to the newest TRACE POINT rather than to the snapshot's `updated`, which is the closer twin
/// of Android reading its own `hrAt`. A publish happens for battery or score changes too, so `updated`
/// refreshes without a new reading and would make a stale heart look fresh.
public enum HrDisplay {
    /// Past this without a newer reading the number is DIMMED: still shown, no longer claimed as live.
    public static let liveWindow: TimeInterval = 2 * 60

    /// Past this it is dropped entirely — too old to stand for the wearer at all.
    public static let staleCap: TimeInterval = 15 * 60

    /// - Returns: the bpm to show (nil to show nothing) and whether it is a carried-over reading.
    public static func resolve(bpm: Int?, newestPointTs: Int64?, now: Date) -> (bpm: Int?, stale: Bool) {
        guard let bpm, bpm > 0 else { return (nil, false) }
        // No trace yet is the FIRST reading, not an old one: a widget added this minute has a bpm and an
        // empty series, and dropping the number there would blank a widget that just started working.
        guard let newestPointTs else { return (bpm, false) }
        let age = now.timeIntervalSince1970 - TimeInterval(newestPointTs)
        if age > staleCap { return (nil, false) }
        return (bpm, age > liveWindow)
    }
}

File: android/app/src/main/java/com/noop/widget/HrTrace.kt
package com.noop.widget

/** One point on the widget's heart-rate trace: when it was taken, and the bpm. */
data class HrPoint(val ts: Long, val bpm: Int)

/**
 * The pure half of the heart-rate widget: what the trace CONTAINS and where its ink goes.
 *
 * Split from the drawing because Glance cannot draw. A sparkline has to be rendered to a Bitmap and
 * handed over as an Image, and a Bitmap is not something a JVM test can meaningfully assert about — so
 * everything decidable without a Canvas lives here, unit-tested, and the renderer is left with nothing
 * but `moveTo`/`lineTo` over coordinates this file produced.
 *
 * Also the shape an iOS twin would share. SwiftUI can draw the trace directly, so the renderer will not
 * port, but the retention rule, the tick choices and the normalisation are the same decisions on both
 * platforms and should not be made twice.
 */
object HrTrace {

    /** How much history the trace shows. Three hours matches what fits legibly at widget width. */
    const val WINDOW_SEC: Long = 3 * 60 * 60

    /** One point per minute. The strap streams ~1/s; keeping every sample would put 10 800 points behind
     *  a trace 300 pixels wide, which costs prefs space and buys nothing the eye can resolve. */
    const val BUCKET_SEC: Long = 60

    /** What counts as a GAP. Deliberately NOT the Today chart's "more than one bucket": that chart reads
     *  a fixed DB grid, where a missing bucket really is missing data, while this series only gains a
     *  point when the app PUBLISHES one, and background scheduling routinely skips a minute. At a
     *  one-bucket threshold that ordinary jitter would shatter a healthy trace into dots, which is a
     *  worse lie than the joined line being fixed.
     *
     *  Set to [HrDisplay.STALE_CAP_MS] rather than to a number picked here, so the line breaks exactly
     *  where the widget would already have dropped the headline reading for being too old to represent
     *  HR at all. `theGapThresholdMatchesTheStaleCap` pins the two together. */
    const val GAP_SEC: Long = 15 * BUCKET_SEC

    /** Hard cap, so a clock jump backwards cannot grow the series without bound. WINDOW_SEC/BUCKET_SEC
     *  is 180; the slack absorbs a boundary point without letting the list run away. */
    const val MAX_POINTS: Int = 200

    /**
     * Fold a fresh reading into the series: one point per minute bucket, newest wins within a bucket.
     *
     * Newest-wins rather than an average because the widget's headline number is the LATEST bpm, and a
     * trace whose last point disagreed with the number printed above it would read as a bug. Within a
     * minute the difference is a beat or two.
     */
    fun append(series: List<HrPoint>, ts: Long, bpm: Int, nowSec: Long = ts): List<HrPoint> {
        if (bpm <= 0) return prune(series, nowSec)
        val bucket = ts / BUCKET_SEC * BUCKET_SEC
        val out = ArrayList<HrPoint>(series.size + 1)
        for (p in series) if (p.ts != bucket) out.add(p)
        out.add(HrPoint(bucket, bpm))
        // No sort here: [prune] sorts on the way out, and it has to anyway because [decode] feeds it
        // untrusted input. Appending at the end is already in order for every sample that is not a clock
        // jump backwards, and prune's sort is what makes even that case correct.
        return prune(out, nowSec)
    }

    /** Drop anything older than the window, then anything beyond the cap (oldest first). */
    fun prune(series: List<HrPoint>, nowSec: Long): List<HrPoint> {
        val floor = nowSec - WINDOW_SEC
        val kept = series.filter { it.ts >= floor }.sortedBy { it.ts }
        return if (kept.size <= MAX_POINTS) kept else kept.subList(kept.size - MAX_POINTS, kept.size)
    }

    /** `ts:bpm` pairs, comma separated. Compact enough that 200 points sit well inside a prefs string,
     *  and readable in a bug report, which a binary blob would not be. */
    fun encode(series: List<HrPoint>): String =
        series.joinToString(",") { "${it.ts}:${it.bpm}" }

    /** Tolerant by design: a malformed pair is skipped rather than throwing. This runs on the widget's
     *  render path, where the alternative to a partial trace is a crashed home screen. */
    fun decode(s: String?): List<HrPoint> {
        if (s.isNullOrBlank()) return emptyList()
        val out = ArrayList<HrPoint>()
        for (part in s.split(',')) {
            val i = part.indexOf(':')
            if (i <= 0 || i == part.length - 1) continue
            val ts = part.substring(0, i).toLongOrNull() ?: continue
            val bpm = part.substring(i + 1).toIntOrNull() ?: continue
            if (bpm > 0) out.add(HrPoint(ts, bpm))
        }
        out.sortBy { it.ts }
        return out
    }

    /** The three numbers the header shows. Null when there is nothing to describe. */
    data class Stats(val min: Int, val max: Int, val latest: Int)

    fun stats(series: List<HrPoint>): Stats? {
        if (series.isEmpty()) return null
        var lo = series[0].bpm
        var hi = series[0].bpm
        for (p in series) {
            if (p.bpm < lo) lo = p.bpm
            if (p.bpm > hi) hi = p.bpm
        }
        return Stats(min = lo, max = hi, latest = series.last().bpm)
    }


    /**
     * The largest drawing box that fits the widget payload budget, keeping the requested aspect.
     *
     * RemoteViews travels to the launcher over a BINDER TRANSACTION with a hard size ceiling, and a
     * widget that exceeds it does not degrade — it fails to render at all. An ARGB_8888 bitmap costs
     * four bytes a pixel, so a 300dp-wide chart on a 4x screen is already about a megabyte on its own,
     * before the rest of the RemoteViews. Dimension caps alone do not express that: the constraint is
     * AREA, and a wide-and-short box and a tall-and-narrow one can both be legal while their product
     * is not.
     *
     * Downscaling is nearly free here in a way it would not be for text or an icon: a sparkline is a
     * smooth line, and the `Image` stretches the result back to the same slot, so the only cost is a
     * fractionally softer stroke.
     *
     * Returned as the box BOTH [points] and the renderer must use. They used to disagree — geometry ran
     * at the requested width while the renderer clamped its own — which silently clipped the right-hand
     * end of the trace on any screen large enough to hit the cap.
     */
    fun fitBox(widthPx: Int, heightPx: Int, maxBytes: Int = MAX_BITMAP_BYTES): Pair<Int, Int> {
        val w = widthPx.coerceAtLeast(1)
        val h = heightPx.coerceAtLeast(1)
        val bytes = w.toLong() * h.toLong() * BYTES_PER_PIXEL
        if (bytes <= maxBytes) return w to h
        val scale = Math.sqrt(maxBytes.toDouble() / bytes.toDouble())
        return (w * scale).toInt().coerceAtLeast(1) to (h * scale).toInt().coerceAtLeast(1)
    }

    /** The bitmap's byte budget. Half a megabyte leaves the rest of the RemoteViews comfortable inside
     *  the transaction ceiling, and still affords a full-density chart on an ordinary phone. */
    const val MAX_BITMAP_BYTES: Int = 512 * 1024

    /** Bytes per pixel the renderer actually spends. The trace is one hue over an opaque card, so it is
     *  drawn RGB_565 rather than ARGB_8888 — there is no transparency to preserve, and at four bytes a
     *  pixel the budget could not afford both a taller chart and a width that avoids stretching it. */
    const val BYTES_PER_PIXEL: Int = 2

    /**
     * The widest bitmap the budget allows at an EXACT height, with headroom over what the caller asked.
     *
     * Two things force this. `LocalSize` under-reports on some launchers — a One UI card reported about
     * 60% of its true width — so a bitmap rendered at the reported width is UPSCALED to fill, and only
     * horizontally, which turns a round stroke elliptical and soft. And [fitBox] preserves aspect, so
     * simply asking for a wider box would shrink the height too and trade a horizontal stretch for a
     * vertical one.
     *
     * So: height is taken as given and never scaled, and the width gets what is left of the budget, up
     * to [WIDTH_HEADROOM] times the request so a small widget does not allocate a huge strip for
     * nothing. Downscaling is the cheap direction — a smooth line loses nothing to it.
     */
    fun widestAtHeight(requestedPx: Int, heightPx: Int, maxBytes: Int = MAX_BITMAP_BYTES): Int {
        val h = heightPx.coerceAtLeast(1)
        val budgetWidth = (maxBytes / (h * BYTES_PER_PIXEL)).coerceAtLeast(1)
        val want = (requestedPx.coerceAtLeast(1) * WIDTH_HEADROOM).toInt()
        return want.coerceAtMost(budgetWidth).coerceAtLeast(requestedPx.coerceAtLeast(1).coerceAtMost(budgetWidth))
    }

    /**
     * How much wider than the reported width to draw, to cover a launcher that under-reports.
     *
     * NOTE THAT THIS CEILING IS RARELY THE ONE THAT BINDS. [MAX_BITMAP_BYTES] runs out first at any
     * realistic widget size: at 420dpi a 300dp-wide card gets 1.66x, a 380dp one 1.26x, and at 480dpi
     * a 380dp card gets 0.96x — narrower than the reported width, so the bitmap is UPSCALED, which is
     * the artefact this constant exists to prevent. Raising this number changes nothing on a real
     * phone; the byte budget is the lever, and it trades directly against horizontal resolution.
     *
     * Kept at two because it is the correct intent and it still binds on a small widget. The gap
     * between the intent and what the budget allows is measured rather than papered over — see the
     * widget cost counters.
     */
    const val WIDTH_HEADROOM: Float = 2f

    /** A point in the trace's pixel box, origin top-left, as the renderer wants it.
     *
     *  [startsRun] means LIFT THE PEN before this point: nothing was recorded for [GAP_SEC] before it.
     *  The renderer joined every point unconditionally, and because x is mapped by TIME rather than by
     *  index, a 90-minute disconnect inside the 3-hour window drew as one confident diagonal across
     *  half the widget. The gap was already the right width; it was the line across it that was never
     *  measured: the same defect #2082 describes on the Today sparkline, on a second surface. */
    data class Pt(val x: Float, val y: Float, val startsRun: Boolean = false)

    /** The trace split into runs of consecutive readings: each range is a stretch the strap recorded
     *  without a break, and the pen lifts between one run and the next. A series with no gaps is one
     *  run, which is the common case and draws exactly as it always did. */
    fun runs(points: List<Pt>): List<IntRange> {
        if (points.isEmpty()) return emptyList()
        val out = mutableListOf<IntRange>()
        var start = 0
        for (i in 1 until points.size) {
            if (points[i].startsRun) {
                out.add(start..(i - 1))
                start = i
            }
        }
        out.add(start..(points.size - 1))
        return out
    }

    /**
     * The trace as pixels inside a `width` x `height` box.
     *
     * Two degenerate shapes decide most of this, and both are ordinary rather than exotic — a widget
     * placed mid-afternoon has one point, and a resting arm holds one bpm for minutes at a time:
     *
     *  - ONE point, or every point at the same instant: there is no time axis to spread across, so it
     *    sits at the left edge rather than at x=NaN.
     *  - every bpm equal: there is no range to scale into, so the line runs along the VERTICAL MIDDLE.
     *    Pinning it to the top or bottom would read as a maxed-out or flatlined heart.
     *
     * Y is flipped on the way out: bpm rises upward, pixels rise downward.
     */
    fun points(series: List<HrPoint>, width: Float, height: Float): List<Pt> {
        if (series.isEmpty() || width <= 0f || height <= 0f) return emptyList()
        val t0 = series.first().ts
        val t1 = series.last().ts
        val span = (t1 - t0).toFloat()
        var lo = series[0].bpm
        var hi = series[0].bpm
        for (p in series) {
            if (p.bpm < lo) lo = p.bpm
            if (p.bpm > hi) hi = p.bpm
        }
        val range = (hi - lo).toFloat()
        return series.mapIndexed { i, p ->
            val x = if (span <= 0f) 0f else (p.ts - t0) / span * width
            val y = if (range <= 0f) height / 2f else height - (p.bpm - lo) / range * height
            Pt(x, y, startsRun = i > 0 && p.ts - series[i - 1].ts > GAP_SEC)
        }
    }

    /**
     * The three bpm labels down the right edge: max, midpoint, min, exactly as they are stacked in the
     * chart. Returned high-to-low so the caller can lay them out top-down without re-sorting.
     *
     * A flat series returns the same number three times rather than an invented spread — the honest
     * reading of a steady heart is one number, and manufacturing 58/59/60 around it would put ticks on
     * the chart that no sample supports.
     */
    fun bpmTicks(stats: Stats): List<Int> =
        listOf(stats.max, (stats.min + stats.max + 1) / 2, stats.min)

    /**
     * The three timestamps along the bottom: first, middle, last of the DATA, not of the window.
     *
     * Anchored to the data because a widget holding forty minutes of history would otherwise label its
     * axis with two hours that were never sampled, and the trace would appear crushed into the right
     * third of a mostly empty chart. Fewer than three distinct instants returns what there is, so the
     * renderer draws one label rather than three copies of it.
     */
    fun timeTicks(series: List<HrPoint>): List<Long> {
        if (series.isEmpty()) return emptyList()
        val first = series.first().ts
        val last = series.last().ts
        if (first == last) return listOf(first)
        val mid = first + (last - first) / 2
        return if (mid == first || mid == last) listOf(first, last) else listOf(first, mid, last)
    }
}
